File: android/app/src/main/kotlin/com/apppostit/apppostit/ime/SqliteReader.kt
package com.apppostit.apppostit.ime

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.util.Log
import java.io.File

private const val TAG = "AppPostItIME"

/**
 * Reads the same SQLite file the main Flutter app (drift) writes to.
 * Table/column names are kept in sync with lib/data/tables.dart by hand --
 * explicit column lists here so a future Dart schema change fails loudly
 * rather than silently misreading.
 */
private const val DATABASE_FILE_NAME = "apppostit.sqlite"

data class CategoryRow(val id: Long, val name: String, val sortOrder: Long)

data class PostRow(
    val id: Long,
    val categoryId: Long,
    val label: String?,
    val body: String,
    val sortOrder: Long,
)

class SqliteReader(private val context: Context) {
    // path_provider's getApplicationDocumentsDirectory() resolves to
    // <app data dir>/app_flutter on Android -- a sibling of filesDir, not
    // nested inside it -- confirmed by inspecting the installed app's
    // private storage on device via `adb shell run-as ... ls`.
    private val dbFile: File =
        File(File(context.filesDir.parentFile, "app_flutter"), DATABASE_FILE_NAME)

    fun getCategories(): List<CategoryRow> {
        return withDb { db ->
            db.rawQuery(
                "SELECT id, name, sort_order FROM categories ORDER BY sort_order",
                null,
            ).use { cursor ->
                val result = mutableListOf<CategoryRow>()
                while (cursor.moveToNext()) {
                    result.add(
                        CategoryRow(
                            id = cursor.getLong(0),
                            name = cursor.getString(1),
                            sortOrder = cursor.getLong(2),
                        )
                    )
                }
                result
            }
        } ?: emptyList()
    }

    fun getPosts(categoryId: Long): List<PostRow> {
        return withDb { db ->
            db.rawQuery(
                "SELECT id, category_id, label, body, sort_order FROM posts " +
                    "WHERE category_id = ? ORDER BY sort_order",
                arrayOf(categoryId.toString()),
            ).use { cursor ->
                val result = mutableListOf<PostRow>()
                while (cursor.moveToNext()) {
                    result.add(
                        PostRow(
                            id = cursor.getLong(0),
                            categoryId = cursor.getLong(1),
                            label = if (cursor.isNull(2)) null else cursor.getString(2),
                            body = cursor.getString(3),
                            sortOrder = cursor.getLong(4),
                        )
                    )
                }
                result
            }
        } ?: emptyList()
    }

    private fun <T> withDb(block: (SQLiteDatabase) -> T): T? {
        if (!dbFile.exists()) return null
        val db = try {
            SQLiteDatabase.openDatabase(dbFile.absolutePath, null, SQLiteDatabase.OPEN_READONLY)
        } catch (e: Exception) {
            Log.e(TAG, "failed to open db", e)
            return null
        }
        return try {
            block(db)
        } catch (e: Exception) {
            Log.e(TAG, "query failed", e)
            null
        } finally {
            db.close()
        }
    }
}

File: ios/AppPostItKeyboard/SqliteReader.swift
import Foundation
import SQLite3

/// Reads the same SQLite file the main Flutter app (drift) writes to, out
/// of the shared App Group container (see AppGroupPlugin.swift /
/// lib/data/database.dart -- the extension can't see the app's own
/// documents directory at all, only the App Group container).
/// Table/column names are kept in sync with lib/data/tables.dart by hand --
/// explicit column lists here so a future Dart schema change fails loudly
/// rather than silently misreading, matching SqliteReader.kt on Android.
private let appGroupId = "group.com.apppostit.apppostit"
private let databaseFileName = "apppostit.sqlite"

struct CategoryRow {
    let id: Int64
    let name: String
    let sortOrder: Int64
}

struct PostRow {
    let id: Int64
    let categoryId: Int64
    let label: String?
    let body: String
    let sortOrder: Int64
}

final class SqliteReader {
    private var dbPath: String? {
        guard let container = FileManager.default.containerURL(
            forSecurityApplicationGroupIdentifier: appGroupId
        ) else { return nil }
        return container.appendingPathComponent(databaseFileName).path
    }

    func getCategories() -> [CategoryRow] {
        return withDb { db in
            var rows: [CategoryRow] = []
            var stmt: OpaquePointer?
            let sql = "SELECT id, name, sort_order FROM categories ORDER BY sort_order"
            guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return rows }
            defer { sqlite3_finalize(stmt) }
            while sqlite3_step(stmt) == SQLITE_ROW {
                rows.append(
                    CategoryRow(
                        id: sqlite3_column_int64(stmt, 0),
                        name: String(cString: sqlite3_column_text(stmt, 1)),
                        sortOrder: sqlite3_column_int64(stmt, 2)
                    )
                )
            }
            return rows
        } ?? []
    }

    func getPosts(categoryId: Int64) -> [PostRow] {
        return withDb { db in
            var rows: [PostRow] = []
            var stmt: OpaquePointer?
            let sql = """
                SELECT id, category_id, label, body, sort_order FROM posts \
                WHERE category_id = ? ORDER BY sort_order
                """
            guard sqlite3_prepare_v2(db, sql, -1, &stmt, nil) == SQLITE_OK else { return rows }
            defer { sqlite3_finalize(stmt) }
            sqlite3_bind_int64(stmt, 1, categoryId)
            while sqlite3_step(stmt) == SQLITE_ROW {
                let label = sqlite3_column_text(stmt, 2).map { String(cString: $0) }
                rows.append(
                    PostRow(
                        id: sqlite3_column_int64(stmt, 0),
                        categoryId: sqlite3_column_int64(stmt, 1),
                        label: label,
                        body: String(cString: sqlite3_column_text(stmt, 3)),
                        sortOrder: sqlite3_column_int64(stmt, 4)
                    )
                )
            }
            return rows
        } ?? []
    }

    private func withDb<T>(_ block: (OpaquePointer) -> T) -> T? {
        guard let path = dbPath, FileManager.default.fileExists(atPath: path) else { return nil }
        var db: OpaquePointer?
        guard sqlite3_open_v2(path, &db, SQLITE_OPEN_READONLY, nil) == SQLITE_OK, let db else {
            sqlite3_close(db)
            return nil
        }
        defer { sqlite3_close(db) }
        return block(db)
    }
}
